### PR TITLE
Create, Delete and Increase Partitions of Message Type

### PR DESCRIPTION
**Pull request by selvakumaar5496**

### CODE REVIEW

File: py-utils/src/utils/message_bus/__init__.py [hunk 1]
```diff
@@ -18,6 +18,6 @@
 __title__ = 'message_bus'
 
 from cortx.utils.message_bus.message_bus import MessageBus
-from cortx.utils.message_bus.message_bus_client import MessageProducer, MessageConsumer
+from cortx.utils.message_bus.message_bus_client import MessageProducer, MessageConsumer, MessageBusAdmin
```
Comment (sachinpunadikar) [inline py-utils/src/utils/message_bus/__init__.py:21]: Remove this from this place, and add it at appropriate place.

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 2]
```diff
@@ -20,7 +20,8 @@
 import json
 import re
 from confluent_kafka import Producer, Consumer
-from confluent_kafka.admin import AdminClient, ConfigResource
+from confluent_kafka.admin import AdminClient, ConfigResource, NewTopic, \
```
Comment (sachinpunadikar) [inline py-utils/src/utils/message_bus/message_broker_collection.py:23]: Remove this from here and add it at appropriate place.

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
Make required changes.
Please note that there will not be any dependent python packages installed, when cortx-py-utils get installed. A quick test would be to install the rpm on fresh VM and run post_install phase of utils_setup

File: py-utils/src/utils/message_bus/__init__.py (hunk 1 above)
Comment (sachinpunadikar) [reply py-utils/src/utils/message_bus/__init__.py:21]: Its okay to have it here.

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 2 above)
Comment (sachinpunadikar) [reply py-utils/src/utils/message_bus/message_broker_collection.py:23]: Its okay to have it here

**Review by sachinpunadikar — APPROVED** [review]
+1

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 3]
```diff
@@ -102,6 +105,84 @@ def init_client(self, client_type: str, **client_conf: dict):
...
+    def _get_metadata(self, admin: object):
+        """ To get the metadata information of message type """
+        try:
+            message_type_metadata = admin.list_topics().__dict__
+            return message_type_metadata['topics']
+        except Exception as e:
+            raise MessageBusError(errno.EINVAL, "Unable to list message type. \
+                %s", e)
+
+    def register_message_type(self, admin_id: str, message_type: list, \
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:126]: ```def register_message_type(self, admin_id: str, message_types: list, \```

Reflect the variable names correctly as per the types. There are multiple message types.

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 4]
```diff
@@ -102,6 +105,84 @@ def init_client(self, client_type: str, **client_conf: dict):
...
+                    list(self._get_metadata(admin).keys()):
+                    if list_retry > self._max_list_message_type_count:
+                        raise MessageBusError(errno.EINVAL, "Maximum retries \
+                            exceeded for creating %s.", each_message_type)
+                    time.sleep(list_retry*1)
+                    continue
+                else:
+                    break
+
+    def deregister_message_type(self, admin_id: str, message_type: list):
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:147]: Same here

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 5]
```diff
@@ -102,6 +105,84 @@ def init_client(self, client_type: str, **client_conf: dict):
...
+                if each_message_type in list(self._get_metadata(admin).keys()):
+                    if list_retry > self._max_list_message_type_count:
+                        raise MessageBusError(errno.EINVAL, "Maximum retries \
+                            exceeded for deleting %s.", each_message_type)
+                    time.sleep(list_retry*1)
+                    continue
+                else:
+                    break
+
+    def increase_parallelism(self, admin_id: str, message_type: list, \
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:164]: Change the name to set_... as you are specying a fixed value. 
```def set_parallelism(self, admin_id: str, message_types: list, \```

Increase would mean increase by N, e.g. +1

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 6]
```diff
@@ -102,6 +105,84 @@ def init_client(self, client_type: str, **client_conf: dict):
...
+        partitions: int):
+        """ Increases the partitions for message type """
+        admin = self._clients['admin'][admin_id]
+        new_partition = [NewPartitions(each_message_type, \
+            new_total_count=partitions) for each_message_type in message_type]
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:169]: What if the new count is less than the old count ? Should we report an error ?

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 7]
```diff
@@ -33,6 +33,45 @@ def _get_conf(self, key: str):
...
+        self._message_bus.deregister_message_type(client_id, message_type)
+
+    def increase_parallelism(self, message_type: list, partitions: int):
+        """
+        To increase the number of partitions of a Message Type
+
+        Parameters:
+        message_type     This is essentially equivalent to the queue/topic name.
+                         For e.g. ["Alert"]
+        partitions       Integer that represents number of partitions to be
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_bus_client.py:68]: What about decrease ?

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 7 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_bus_client.py:68]: Number of partitions for a topic can only be increased, never decreased

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 6 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:169]: yes, Kafka throws an error if the number of partitions is less than the existing number

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 5 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:164]: yes, we can only increase by N

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 3 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py]: changed as suggested

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 4 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py]: changed as suggested

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 6 above)
Comment (ujjwalpl) [reply py-utils/src/utils/message_bus/message_broker_collection.py]: We need to catch and throw a Message Bus Error. Can you confirm if this is the case ?

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 7 above)
Comment (ujjwalpl) [reply py-utils/src/utils/message_bus/message_bus_client.py:68]: ```def increase_parallelism(self, message_types: list, partitions: int):```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 7 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_bus_client.py:68]: taken care

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 6 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py]: This exception is caught in **def _task_status(self, tasks :dict):** method

File: py-utils/src/utils/message_bus/message_bus_client.py [hunk 8]
```diff
@@ -87,21 +139,21 @@ class MessageConsumer(MessageBusClient):
...
         auto_ack        Can be set to "True" or "False"
         offset          Can be set to "earliest" (default) or "latest".
                         ("earliest" will cause messages to be read from the
                         beginning)
         """
         super().__init__(message_bus, client_type='consumer', \
             client_id=consumer_id, consumer_group=consumer_group, \
-            message_type=message_type, auto_ack=auto_ack, offset=offset)
\ No newline at end of file
+            message_type=message_types, auto_ack=auto_ack, offset=offset)
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_bus_client.py:159]: List is assigned to a str.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py [hunk 9]
```diff
@@ -67,12 +66,16 @@ def init_client(self, client_type: str, **client_conf: dict):
...
+            admin = AdminClient(kafka_conf)
+            self._clients['admin'][client_conf['client_id']] = admin
+
         if client_type == 'producer':
             producer = Producer(**kafka_conf)
             self._clients[client_type][client_conf['client_id']] = producer
 
             self._resource = ConfigResource('topic', client_conf['message_type'])
-            conf = self._admin.describe_configs([self._resource])
+            conf = admin.describe_configs([self._resource])
```
Comment (ujjwalpl) [inline py-utils/src/utils/message_bus/message_broker_collection.py:78]: Please note client_conf['message_type'] has multiple message types. Change the key to reflect the same, i.e. "message_types"

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/message_bus/message_broker_collection.py (hunk 9 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_broker_collection.py:78]: changed for consumer, since it takes list of message_types

**Review by ujjwalpl — APPROVED** [review]

File: py-utils/src/utils/message_bus/message_bus_client.py (hunk 8 above)
Comment (selvakumaar5496) [reply py-utils/src/utils/message_bus/message_bus_client.py]: taken care

**Pull request merged**